slack_today_i_did/parser.py

```diff
@@ -1,3 +1,24 @@
+from typing import Any, TypeVar, Callable, List, NamedTuple, Union
+import copy
+import functools
+
+
+FuncArg = Union['Constant', 'FuncCall']
+Constant = NamedTuple('Constant', [('value', Any), ('return_type', type)])
+FuncCall = NamedTuple('FuncCall', [('func_name', str), ('args', List[FuncArg]), ('return_type', type)])
+FuncResult = NamedTuple(
+    'FuncResult',
+    [('result', Any), ('action', Callable), ('args', List[Any]), ('errors', List[str])])
+
+
+def metafunc(fn):
+    fn.is_metafunc = True
+    return fn
+
+
+def is_metafunc(fn):
```

Comment:
Right - I meant making it so that the parser can basically pattern match on the type of function. E.g 

```
case func of 
  MetaFunc fn -> ..
  Func fn -> ..
  Constant const -> ..
```